service/controller/v15/resource/drainfinisher/create.go

```diff
@@ -0,0 +1,162 @@
+package drainfinisher
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	corev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/aws-operator/service/controller/v15/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/v15/key"
+)
+
+// EnsureCreated tries to drain guest cluster nodes when necessary. Once it
+// detects a guest cluster node being in terminating/wait state in EC2 a
+// DrainerConfig is created to instruct the node-operator to drain the specific
+// node. The node-operator updates the DrainerConfig state as soon as it has
+// drained the node and the aws-operator here completes the lifecycle hook of
+// the drained node and deletes the DrainerConfig.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	controllerCtx, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	workerASGName := controllerCtx.Status.Drainer.WorkerASGName
+	if workerASGName == "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "worker ASG name is not available yet")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	var drainedDrainerConfigs []corev1alpha1.DrainerConfig
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding out drained drainer configs for the guest cluster")
+
+		n := v1.NamespaceAll
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(customObject)),
+		}
+
+		drainerConfigs, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(n).List(o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, drainerConfig := range drainerConfigs.Items {
+			if drainerConfig.Status.HasDrainedCondition() {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("drainer config '%s' of guest cluster has drained condition", drainerConfig.GetName()))
+				drainedDrainerConfigs = append(drainedDrainerConfigs, drainerConfig)
+			}
+
+			if drainerConfig.Status.HasTimeoutCondition() {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("drainer config '%s' of guest cluster has timeout condition", drainerConfig.GetName()))
+				drainedDrainerConfigs = append(drainedDrainerConfigs, drainerConfig)
+			}
+		}
+
+		if len(drainedDrainerConfigs) == 0 {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find drained drainer configs for the guest cluster")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "nothing to do")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
```

Comment:
When resources are split and have single responsibility we can cancel early. `nothing to do` is just a reason log for those who are not familiar with the codebase, and for those who forget quickly (like myself).